Service/ExecuteSourceDeductionForItems.php

```diff
@@ -0,0 +1,153 @@
+use Magento\Catalog\Model\Indexer\Product\Price\Processor;
+use Magento\Sales\Model\Order\Item as OrderItem;
+use Magento\Sales\Model\Order\Creditmemo\Item as CreditmemoItem;
+
+/**
+ * Class ExecuteSourceDeductionForItems
+ * @package Ampersand\DisableStockReservation\Service
+ */
+class ExecuteSourceDeductionForItems
+{
+    /**
+     * @var Processor
+     */
+    private $priceIndexer;
+
+    /**
+     * @var WebsiteRepositoryInterface
+     */
+    private $websiteRepository;
+
+    /**
+     * @var SalesChannelInterfaceFactory
+     */
+    private $salesChannelFactory;
+
+    /**
+     * @var SalesEventInterfaceFactory
+     */
+    private $salesEventFactory;
+
+    /**
+     * @var ItemToDeductFactory
+     */
+    private $itemToDeductFactory;
+
+    /**
+     * @var SourceDeductionRequestFactory
+     */
+    private $sourceDeductionRequestFactory;
+
+    /**
+     * @var SourceDeductionService
+     */
+    private $sourceDeductionService;
+
+    /**
+     * @var SourcesRepositoryInterface
+     */
+    private $sourceRepository;
+
+    /**
+     * ExecuteSourceDeductionForItems constructor.
+     * @param WebsiteRepositoryInterface $websiteRepository
+     * @param SalesEventInterfaceFactory $salesEventFactory
+     * @param SalesChannelInterfaceFactory $salesChannelFactory
+     * @param ItemToDeductFactory $itemToDeductFactory
+     * @param SourceDeductionRequestFactory $sourceDeductionRequestFactory
+     * @param SourceDeductionService $sourceDeductionService
+     * @param SourcesRepositoryInterface $sourceRepository
+     * @param Processor $priceIndexer
+     */
+    public function __construct(
+        WebsiteRepositoryInterface $websiteRepository,
+        SalesEventInterfaceFactory $salesEventFactory,
+        SalesChannelInterfaceFactory $salesChannelFactory,
+        ItemToDeductFactory $itemToDeductFactory,
+        SourceDeductionRequestFactory $sourceDeductionRequestFactory,
+        SourceDeductionService $sourceDeductionService,
+        SourcesRepositoryInterface $sourceRepository,
+        Processor $priceIndexer
+    ) {
+        $this->websiteRepository = $websiteRepository;
+        $this->salesEventFactory = $salesEventFactory;
+        $this->salesChannelFactory = $salesChannelFactory;
+        $this->itemToDeductFactory = $itemToDeductFactory;
+        $this->sourceDeductionRequestFactory = $sourceDeductionRequestFactory;
+        $this->sourceDeductionService = $sourceDeductionService;
+        $this->sourceRepository = $sourceRepository;
+        $this->priceIndexer = $priceIndexer;
+    }
+
+    /**
+     * @param OrderInterface $order
+     * @param array $itemsToCancel
+     * @param Creditmemo|null $creditmemo
+     */
+    public function executeSourceDeductionForItems(OrderInterface $order, array $itemsToCancel, Creditmemo $creditmemo = null)
+    {
+        $type = SalesEventInterface::EVENT_ORDER_CANCELED;
+        if ($creditmemo) {
```

Comment:
Wouldn't it make more sense to pass this as a flag? We're not using any of the functionality of a "credit memo" here